packages/ui/client/composables/explorer/search.ts

```diff
@@ -0,0 +1,107 @@
+import type { Ref } from 'vue'
+import {
+  filter,
+  filteredFiles,
+  initialized,
+  isFiltered,
+  isFilteredByStatus,
+  openedTreeItems,
+  search,
+  testsTotal,
+  treeFilter,
+  uiEntries,
+} from './state'
+import { explorerTree } from '~/composables/explorer'
+
+export function useSearch(searchBox: Ref<HTMLDivElement | undefined>) {
+  const disableFilter = computed(() => {
+    if (isFilteredByStatus.value) {
+      return false
+    }
+
+    return !filter.onlyTests
+  })
+  const disableClearSearch = computed(() => search.value === '')
+  const debouncedSearch = ref(search.value)
+
+  debouncedWatch(search, (value) => {
+    debouncedSearch.value = value?.trim() ?? ''
+  }, { debounce: 256 })
+
+  function clearSearch(focus: boolean) {
+    search.value = ''
+    focus && searchBox.value?.focus()
+  }
+
+  function clearFilter(focus: boolean) {
+    filter.failed = false
+    filter.success = false
+    filter.skipped = false
+    filter.onlyTests = false
+    focus && searchBox.value?.focus()
+  }
+
+  function clearAll() {
+    clearFilter(false)
+    clearSearch(true)
+  }
+
+  function updateFilterStorage(
+    searchValue: string,
+    failedValue: boolean,
+    successValue: boolean,
+    skippedValue: boolean,
+    onlyTestsValue: boolean,
+  ) {
+    if (!initialized.value) {
+      return
+    }
+
+    treeFilter.value.search = searchValue?.trim() ?? ''
+    treeFilter.value.failed = failedValue
+    treeFilter.value.success = successValue
+    treeFilter.value.skipped = skippedValue
+    treeFilter.value.onlyTests = onlyTestsValue
+  }
+
+  watch(
+    () => [
+      debouncedSearch.value,
+      filter.failed,
+      filter.success,
+      filter.skipped,
+      filter.onlyTests,
+    ] as const,
+    ([search, failed, success, skipped, onlyTests]) => {
+      updateFilterStorage(search, failed, success, skipped, onlyTests)
+      explorerTree.filterNodes()
+    },
+    { flush: 'post' },
+  )
+
+  watch(() => openedTreeItems.value.length, (size) => {
+    if (size) {
+      treeFilter.value.expandAll = undefined
+    }
+  }, { flush: 'post' })
+
+  onMounted(() => {
+    nextTick(() => (initialized.value = true))
```

Comment:
we should change initialized on `resumeRun` or `endRun` in the tree.ts module: initialized is for page refresh, check `Explorer.vue` (we also need to add some spinned with `v-else` L143)